// coffeekraken-compile-server/server/server.js
const __express = require('express');
const __path = require('path');
const __fs = require('fs');
const __nodeSass = require('node-sass');
const __webpack = require('webpack');
const __cors = require('cors');
const __bodyParser = require('body-parser');
const __extend = require('lodash/extend');
const __shortid = require('shortid');
const __exec = require('child_process').exec;
const __HtmlEntities = require('html-entities').AllHtmlEntities;
const __stylus = require('stylus');
const __Cryptr = require('cryptr');

module.exports = function(config) {

	// creating the app
	const app = __express();

	// cryptr instance
	let cryptr;
	if (config.secret) {
		cryptr = new __Cryptr(config.secret);
	}

	// parser body
	app.use(__bodyParser.json());
	app.use(__bodyParser.urlencoded({ extended: true }));

	// cors
	app.use(__cors());

	// attach config to request
	app.use((req, res, next) => {
		req.config = Object.assign({}, config);
		next();
	});

	// pwd
	app.use((req, res, next) => {
		// environment
		let pwd;
		if (req.query.cwd) {
			if (cryptr) {
				pwd = cryptr.decrypt(req.query.cwd);
			} else {
				pwd = req.query.cwd;
			}
		} else {
			pwd = process.env.PWD;
		}
		// set pwd in config
		req.config.pwd = pwd;

		// next
		next();
	});

	// Stylus
	app.post(/\/compile\/(stylus|styl)$/, function (req, res) {

		// check code
		if ( ! req.body.data) {
			res.status(500).json({
				error : 'Missing "data" argument'
			});
			return;
		}

		// htmlentities
		const htmlEntities = new __HtmlEntities();
		console.log("hiiii", req.body)
		// options
		const options = __extend({
			compress: true
		}, req.body.options || {});

		// compile using sass
		__stylus(htmlEntities.decode(req.body.data), options)
		.include(req.config.pwd)
		.render((error, css) => {
			if ( ! error) {
				res.status(200).json({
					language : 'css',
					data : css
				});
			} else {
				res.status(500).json({
					error : error
				});
			}
		});
	});

	// SASS
	app.post(/\/compile\/(sass|scss)$/, function (req, res) {

		// check code
		if ( ! req.body.data) {
			res.status(500).json({
				error : 'Missing "data" argument'
			});
			return;
		}

		// htmlentities
		const htmlEntities = new __HtmlEntities();

		// options
		const options = __extend({
			data : htmlEntities.decode(req.body.data),
			outputStyle: 'compressed'
		}, req.body.options || {});
		// include paths
		options.includePaths = [].concat(options.includePaths, req.config.pwd);

		// compile using sass
		__nodeSass.render(options, (error, result) => {
			if ( ! error) {
				res.status(200).json({
					language : 'css',
					data : result.css.toString()
				});
			} else {
				res.status(500).json({
					error : error
				});
			}
		});
	});

	// JS
	app.post(/\/compile\/(js|javascript|coffee|coffeescript|ts|typescript)$/, function (req, res) {

		// check code
		if ( ! req.body.data) {
			res.status(500).json({
				error : 'Missing "data" argument'
			});
			return;
		}

		console.log("hiiii", req.body)

		// file extension
		let ext = '.js';
		if (req.path.match(/\/(ts|typescript)$/)) ext = '.ts';
		if (req.path.match(/\/(coffee|coffeescript)$/)) ext = '.coffee';

		// generate file hash
		const fileName = __shortid.generate() + ext;
		const filePath = req.config.pwd + '/' + fileName;
		const dirPath = __path.dirname(filePath);

		// htmlentities
		const htmlEntities = new __HtmlEntities();

		// make sure .tmp directory exist
		if ( ! __fs.existsSync(dirPath)) {
			__fs.mkdirSync(dirPath);
		}

		// write file content to use
		__fs.writeFileSync(filePath, htmlEntities.decode(req.body.data));

		// options
		const options = __extend({
			context : req.config.pwd,
			entry : filePath,
			target: 'web',
			output : {
				path : dirPath,
				filename : fileName
			},
			module : {
				rules : [{
					test: /\.js$/,
					exclude: /(node_modules|bower_components)/,
					loader: 'babel-loader'
				}, {
					test: /\.coffee$/,
					exclude: /(node_modules|bower_components)/,
					loader: 'coffee-loader'
				}, {
					test: /\.tsx?$/,
					exclude: /(node_modules|bower_components)/,
					loader: 'ts-loader'
				}]
			}
		}, req.body.options || {});

		// compile using sass
		__webpack(options, (error, result) => {
			if ( ! error) {
				const content = __fs.readFileSync(filePath, 'utf8');
				res.status(200).json({
					language : 'js',
					data : content
				});
			} else {
				res.status(500).json({
					error : error
				});
			}
			// delete file
			__exec('rm -rf ' + filePath);
		});
	});

	// start demo server
	app.listen(config.port, function () {
		console.log('Compile Server : ✓ running on port ' + config.port + '!');
	});
}
