// coffeekraken-s-dialog-component/code-playground.config.js
module.exports = {
	// server port
	port : 3000,

	// title
	title : 's-dialog-component',

	// layout
	layout : 'right',

	// compile server
	compileServer : {

		// compile server port
		port : 4000

	},

	// editors
	editors : {
		html : {
			language : 'html',
			data : `
				<div class="container">
					<h1 class="h1 m-b-small">
						Coffeekraken s-dialog-component
					</h1>
					<p class="p m-b">
						Powerful, fully featured and easy to use dialog component
					</p>
					<a href="#my-cool-dialog" class="btn btn--primary">Open dialog</a>
					<div class="hidden">
						<s-dialog id="my-cool-dialog" modal>
							<h1 class="h1 m-b">
								Hello world
							</h1>
							<p class="p m-b">
								Vivamus vestibulum ultrices eros nec bibendum. In sit amet ultrices mi. Etiam nunc ante, efficitur ac aliquam eget, iaculis quis massa. Duis quis molestie orci. Sed ultricies sem ante, in.
							</p>
							<a class="btn btn--secondary" s-dialog-ok>
								Close dialog
							</a>
						</s-dialog>
						<s-dialog opened>
							<h1 class="h1 m-b">
								Automatically opened dialog
							</h1>
							<p class="p">
								Vivamus vestibulum ultrices eros nec bibendum. In sit amet ultrices mi. Etiam nunc ante, efficitur ac aliquam eget, iaculis quis massa. Duis quis molestie orci. Sed ultricies sem ante, in.
							</p>
						</s-dialog>
					</div>
				</div>
			`
		},
		css : {
			language : 'sass',
			data : `
				@import 'node_modules/coffeekraken-sugar/index';
				@include s-init();
				@include s-classes();
				@include s-typography-classes();
				@include s-button-classes();
				@include s-form-classes();
				body {
					background-image: linear-gradient(120deg, #fdfbfb 0%, #ebedee 100%);
				}
				.container {
					@include s-position(absolute, middle, center);
					min-width:80vw;
				}
				@keyframes dialog-in-overlay {
					from { opacity:0; }
					to {opacity:1; }
				}
				@keyframes dialog-out-overlay {
					from { opacity:1; }
					to {opacity:0; }
				}
				@keyframes dialog-in {
					from {
						opacity:0;
						transform: translateY(-50px);
					}
					to {
						opacity:1;
						transform: translateY(0);
					}
				}
				@keyframes dialog-out {
					from {
						opacity:1;
						transform: translateY(0);
					}
					to {
						opacity:0;
						transform: translateY(50px);
					}
				}
				.s-dialog__overlay {
					background: s-color(secondary, -opacity .8);
					animation: dialog-in-overlay .1s ease-in-out 0s;
				}
				.s-dialog__content {
					max-width: 80%;
					background: white;
					padding: s-space(big);
					animation: dialog-in .2s ease-in-out 0s;
				}
				.s-dialog--out {
					.s-dialog__content {
						animation: dialog-out .3s ease-in-out 0s forwards;
					}
					.s-dialog__overlay {
						animation: dialog-out-overlay .3s ease-in-out 0s forwards;
					}
				}
			`
		},
		js : {
			language : 'js',
			data : `
				import 'webcomponents.js/webcomponents-lite'
				import SDialogComponent from './dist/index'
			`
		}
	}
}
